src/algorithms/MPolyEvaluate.jl

```diff
@@ -0,0 +1,289 @@
+  r = -1
+  cur = f
+  f_prev = -1
+  r_prev = -1
+  while cur != -1
+    next = Blist[1+cur]
+    if Bexps[1+cur][1+k] == 0
+      if f_prev == -1
+        f = Blist[1+cur]
+      else
+        Blist[1+f_prev] = Blist[1+cur]
+      end
+      if r_prev == -1
+        r = cur
+      else
+        Blist[1+r_prev] = cur
+      end
+      Blist[1+cur] = -1
+      r_prev = cur
+    else
+      # mdegs[k] should be minimum non zero exponent
+      Bexps[1+cur][1+k] -= mdegs[1+k]
+      @assert Bexps[1+cur][1+k] >= 0
+      f_prev = cur
+    end
+    cur = next
+  end
+  #stack[end].r = r
+  stack[end] = stack_entry(stack[end].f,
+                           r,
+                           stack[end].v_var,
+                           stack[end].v_exp,
+                           stack[end].ret)
+
+  # convert the quotient
+  push!(stack, stack_entry(f, 0, 0, 0, 1))
+  @goto horner_form
+@label horner_form_ret_1
+
+  # convert the remainder
+  r = stack[end].r
+
+  if r != -1
+    rp += 1
+    @assert 0 <= rp <= nvars
+    push!(stack, stack_entry(r, 0, 0, 0, 2))
+    @goto horner_form
+@label horner_form_ret_2
+
+    if rtypes[1+rp-1] == -1 && rtypes[1+rp] == -1
+      # both quotient and remainder are polynomials
+      regs[1+rp-1], temp = mulpow!(regs[1+rp-1], C[1+stack[end].v_var], stack[end].v_exp, temp)
+      temp = add!(temp, regs[1+rp-1], regs[1+rp])
+      regs[1+rp-1], temp = temp, regs[1+rp-1]
+    elseif rtypes[1+rp-1] == -1 && rtypes[1+rp] != -1
+      # quotient is a polynomial, remainder is a scalar
+      regs[1+rp-1], temp = mulpow!(regs[1+rp-1], C[1+stack[end].v_var], stack[end].v_exp, temp)
+      regs[1+rp-1] = add!(regs[1+rp-1], regs[1+rp-1], Bcoeffs[1+rtypes[1+rp]])
+    elseif rtypes[1+rp-1] != -1 && rtypes[1+rp] == -1
+      # quotient is a scalar, remainder is a polynomial
+      temp = pow!(temp, C[1+stack[end].v_var], stack[end].v_exp)
+      temp = mul!(temp, temp, Bcoeffs[1+rtypes[1+rp-1]])
+      regs[1+rp-1] = add!(regs[1+rp-1], temp, regs[1+rp])
+    else
+      # done by simple case
+      @assert false
+    end
+    rp -= 1
+    @assert 0 <= rp <= nvars
+  else
+    # remainder is zero
+    @assert rtypes[1+rp] == -1
+    # quotient is a polynomial
+    regs[1+rp], temp = mulpow!(regs[1+rp], C[1+stack[end].v_var], stack[end].v_exp, temp)
+  end
+  rtypes[1+rp] = -1
+
+@label horner_form_return
+  # two possible recursive call sites
+  ret = stack[end].ret
+  pop!(stack)
+  if ret == 1
+    @goto horner_form_ret_1
+  elseif ret == 2
+    @goto horner_form_ret_2
+  end
+
+  @assert rp == 0
+  @assert length(stack) == 0
+
+  if rtypes[1+rp] == -1
+    return regs[1+rp]
+  else
+    return ctxC(Bcoeffs[1+rtypes[1+rp]])
+  end
+end
+
+function evaluate_horner(B::MPolyElem, C::Vector{<:RingElem})
+  @assert nvars(parent(B)) <= length(C)
+  ctxC = parent(C[1])
+  Bcoeffs = collect(coefficients(B))
+  ctxA = parent(one(coefficient_ring(B)) * one(ctxC))
+  ctxA === ctxC || error("probably not implemented")
```

Comment:
So the code cannot handle fmpq_mpoly evaluating at fmpz's?

Reply:
Currently cannot. Will try.

Reply:
I don't quite understand. You have the generic fallback, so why is it not working?

Reply:
you have to predict the type of the return of a+b. It will be working soon, but requres some changes.

Reply:
ok, thanks

Reply:
do we have any example where a+b lies in neither parent(a) nor parent(b)?

Reply:
Sorry, I missed your message. Yes, I think this is possible.